contracts/cw-token-swap/src/contract.rs

```diff
@@ -0,0 +1,247 @@
+#[cfg(not(feature = "library"))]
+use cosmwasm_std::entry_point;
+use cosmwasm_std::{
+    to_binary, Addr, Binary, Deps, DepsMut, Env, MessageInfo, Response, StdResult, Uint128,
+};
+use cw2::set_contract_version;
+use cw_storage_plus::Item;
+use cw_utils::must_pay;
+
+use crate::{
+    error::ContractError,
+    msg::{ExecuteMsg, InstantiateMsg, QueryMsg, StatusResponse},
+    state::{CheckedCounterparty, CheckedTokenInfo, COUNTERPARTY_ONE, COUNTERPARTY_TWO},
+};
+
+const CONTRACT_NAME: &str = "crates.io:cw-token-swap";
+const CONTRACT_VERSION: &str = env!("CARGO_PKG_VERSION");
+
+#[cfg_attr(not(feature = "library"), entry_point)]
+pub fn instantiate(
+    deps: DepsMut,
+    _env: Env,
+    _info: MessageInfo,
+    msg: InstantiateMsg,
+) -> Result<Response, ContractError> {
+    set_contract_version(deps.storage, CONTRACT_NAME, CONTRACT_VERSION)?;
+
+    let counterparty_one = msg.counterparty_one.into_checked(deps.as_ref())?;
+    let counterparty_two = msg.counterparty_two.into_checked(deps.as_ref())?;
```

Comment:
do we intend to allow the case where the two counterparties are the same address?

Reply:
Probably good to have a check for that.

Reply:
Done. Great catch!